lib/gslab_model/ModelData.R

```diff
@@ -0,0 +1,190 @@
+        .self$var      <- data
+        .self$varnames <- colnames(.self$var)
+        .self$nvars    <- ncol(.self$var)
+        .self$nobs     <- nrow(.self$var)
+    },
+    
+    isVariable = function(varname) {
+        varname %in% .self$varnames
+    },
+    
+    addArrayVars = function(data, name = "arrayvar") {
+        # Input: a matrix or a 3-dimension array, added in the field var as a 3-dimension variable.
+        
+        d <- dim(data)
+        if (is.null(d))
+            stop("Not an array variable. Use addData instead")
+        if (d[1] != .self$nobs)
+            stop("Length of array variable does not match data")
+        if (length(d) > 3)
+            stop("Array structure is too complicated")
+        if (length(d) == 2)
+            data <- array(data, c(d[1], d[2], 1))
+        .self$var[[name]] <- data
+        .self$varnames    <- colnames(.self$var)
+        .self$nvars       <- ncol(.self$var)
+    },
+    
+    expandArrayVars = function() {
+        # Expand all array variables [varname] with dimensions R by C into multiple variables 
+        #   with the variable name [varname]_array_[1:R]_[1:C]
+        # The reverse function of collapseArrayVars
+        
+        vars <- .self$varnames
+        for (varname in vars) {
+            d <- dim(.self$var[[varname]])
+            if (!is.null(d)) {
+                for (i in 1:d[2]) {
+                    for (j in 1:d[3]) {
+                        newname <- paste(varname, "_array_", i, "_", j, sep = "")
+                        .self$addData(.self$var[[varname]][, i, j], names = newname)
+                    }
+                }
+                .self$removeData(varname)
+            }
+        }
+    },
+    
+    collapseArrayVars = function() {
+        # Collapse all variables with the name [varname]_array_[d1]_[d2] to an array variable 
+        #   [varname] with dimensions R = max(d1) by C = max(d2)
+        # The reverse function of expandArrayVars
+        
+        arrays <- grep("_array_[0-9]+_[0-9]+$", .self$varnames, value = TRUE)
+        arrayvars <- unique(sub("_array_[0-9]+_[0-9]+$", "", arrays))
+        for (arrayvar in arrayvars) {
+            f = function (array) sub(paste(arrayvar, "_array_", sep = ""),  "", array)
+            collection <- lapply(grep(arrayvar, arrays, value = TRUE), f)
+            maxDimension <- as.integer(unlist(strsplit(max(unlist(collection)), "_")))
+            newArray <- array(0, c(.self$nobs, maxDimension))
+            for (array in grep(arrayvar, arrays, value = TRUE)) {
+                loc <- as.integer(unlist(strsplit(sub(paste(arrayvar, "_array_", sep = ""), 
+                                                      "", array), "_")))
+                newArray[, loc[1], loc[2]] = .self$var[[array]]
+                .self$removeData(array)
+            }
+            .self$addArrayVars(newArray, name = arrayvar)
+        }
+    },
+    
+    saveToDisk = function(directory, name, precision = 4) {
```

Comment:
4 seems like a low default precision... I understand that this was the MATLAB default, though, so feel free to keep it as it is.